In [1]:
from IPython.display import HTML
HTML('''
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.
js"></script>
<script>
code_show=true;
function code_toggle() {
if (code_show){
$('div.jp-CodeCell > div.jp-Cell-inputWrapper').hide();
} else {
$('div.jp-CodeCell > div.jp-Cell-inputWrapper').show();
}
code_show = !code_show
}
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" 
value="Click here to toggle on/off the raw code."></form>
''')

In [2]:
from IPython.core.display import HTML
HTML("""
<style>
.output_png {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
}
</style>
""")

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Twice</h1>

<center><img src="./images/ML3_Twice_Title.png" width="90%" height="90%"></center>

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Introduction</h1>

Are you new to K-pop and find associating the name with a face difficult? Are you already following your favorite K-pop group but still cannot answer the common question, who is your bias? Because you cannot remember the name. With the help of Artificial Intelligence (AI), we will create a face recognition algorithm to help you associate the name with the face so you spend more time being a fan and supporting them rather than just trying to remember which is which.

We will use Twice as the K-pop group to establish face recognition for our case.

<h2 style="color:#FDB3A5">Background of Twice</h2>

***

**One in a Million Annyeong Haseyo Twice Imnida!**

Twice is a K-pop girl group comprising nine members: Jihyo, Nayeon, Jeongyeon, Momo, Sana, Mina, Dahyun, Chaeyoung, and Tzuyu. Known for their catchy tunes, vibrant visuals, and energetic performances, Twice has garnered a massive global fanbase. Under the production company JYP Entertainment Co., Ltd, the group finds success in the K-pop industry locally and internationally.

The group debuted last October 20, 2015, through a survival show, Sixteen. Now, the group has been performing together for over seven years. They debuted with an extended play (EP) named *The Story Begins* with the lead single *Like Ooh-Ahh*. Since then, the group has released numerous singles and albums that dominated music charts. Their popularity exploded with tracks such as *Cheer Up*, *TT*, *Fancy*, and *What is Love?*. With their fans called *Once*, they supported the group in every release. Now in 2023, Twice are now on its 5th World Tour, still dominating the music industry.

<h2 style="color:#FDB3A5">Motivation</h2>

***

To those new to K-pop, it is difficult to distinguish and memorize the names and associate the name with a face. For me, this task of remembering the name of Twice took months. With the power of Artificial Intelligence (AI), we can create a face recognition for Twice to help new fans identify which member it is.

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Methodology</h1>

<center><img src="./images/ML3_Twice_Methodology.png" width="90%" height="90%"></center>
<center>Image 01. Project Methodology.</center>

**Data Collection**

The images of Twice are downloaded from the gallery of its official website.<sup>[4]</sup> It is manually sorted into a folder containing the name of the member (1 folder for each member). Each image used is just a solo picture meaning only the face of the Twice member can be seen. For this project, we used 20 images for each member (180 total solo images).
<br><br>

**Face Detection**

`MTCNN` was used as the pre-trained model for *Face Detection*. This model will detect the faces within the image and provide a bounding box on the face. The image will then be cropped according to this bounding box to retrieve only the image of the face.
<br><br>

**Face Classification**

The cropped-face images of Twice will be split into 60% Training (12 images), 20% Validation (4 images), and 20% Testing (4 images). `Inception-ResNetV1` pre-trained on `VGGFace2` was used to classify cropped-face images into their respective correct name. *Feature Extraction* approach was used to tailor-fit the model towards the dataset. 

Combining the concept of *Face Detection* and *Face Classification*, we can now determine who is that Twice member? A sample group image was used to illustrate how this would work.

<h2 style="color:#FDB3A5">Deep Learning on Images</h2>

***

The deep learning method, as mentioned above, has two parts: *Face Detection* and *Face Classification*. 

For *Face Detection* I used the model based on the paper by Zhang et al. in 2016 titled [Joint Face Detection and Alignment Using Multitask Cascaded Convolutional Networks](https://arxiv.org/ftp/arxiv/papers/1604/1604.02878.pdf). Shown below is the architecture of the model.

<center><img src="./images/MTCNN_architecture.jpg" width="70%" height="70%"></center>
<center>Image 02. MTCNN Model Architecture.<sup>[1]</sup></center>

The model has three stages:
1. Proposal Network (P-Net): Produces candidate windows and their bounding boxes.
2. Refine Network (R-Net): Rejects many false candidates.
3. Output Network (O-Net): Describes the face in more detail (outputs five facial landmarks' positions).

This pre-trained model no longer needs feature extraction or fine-tuning since I will use it at its intended use, that is, to detect faces (including bounding box and landmark positions).

For Face Classification, the model I used is based on the paper by Szegedy et al. in 2016 titled [Inception-v4, Inception-ResNet and the Impact of Residual Connections on Learning](https://arxiv.org/pdf/1602.07261.pdf). Shown below is the architecture of the model.

<center><img src="./images/InceptionResnetV1_architecture.jpg" width="30%" height="50%"></center>
<center>Image 03. Inception-ResNet-v1 Model Architecture. <sup>[2]</sup></center>

The complete architecture of Inception-ResNet-v1 is shown in its paper. To summarize, here are what each block does in the model.
1. Stem: responsible for extracting basic features of the image (i.e., edge detection and simple patterns).
2. Inception-Resnet: consists of multiple parallel convolutional layers with different kernel sizes to capture local and global features. It also has residual connections to address the vanishing gradient problem.
3. Reduction: downsamples the spatial dimension of the feature map but increases the number of channels.

This *Face Classification* model is pre-trained on VGGFace2, which is based on the paper by Cao et al. in 2016 titled [VGGFace2: A dataset for recognizing faces across pose and age](https://arxiv.org/pdf/1710.08092v2.pdf).<sup>[3]</sup> The dataset contains 3.31 million photos of 9131 subjects. Images are downloads from Google Images, with varying poses, age, illumination, ethnicity, and profession.

<h2 style="color:#FDB3A5">Setting Up</h2>

***

Before everything, we must first install the necessary libraries. We used the following non-standard Python libraries

* `Pillow`, `scikit-image`, `scikit-learn`, `torch`, `torchvision`, `torchsummary`, `mtcnn`, and `facenet-pytorch`.

In [3]:
# pip install Pillow
# pip install scikit-image
# pip install scikit-learn
# pip install torch
# pip install torchvision
# pip install torchsummary
# pip install mtcnn
# pip install facenet-pytorch

After installing the non-standard Python libraries, let us import all necessary libraries.

**Basic Imports:** `os`, `sys`, `time`, `copy`, `shutil`, `zipfile`, `numpy`, and `matplotlib`.

**Image Processing:** `PIL`, and `skimage`.

**Deep Learning:** `scikit-learn`, `torch`, `torchvision`, `torchsummary`, `mtcnn`, and `facenet_pytorch`.

In [4]:
# Quick fix for permission error
import os
os.environ['SKIMAGE_DATADIR'] = '/tmp/.skimage_cache'
os.environ['XDG_CACHE_HOME'] = '/home/msds2023/fgarcia/.cache'

In [29]:
# Basic Imports
import sys
import time
import copy
import shutil
import zipfile
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle, Circle

# Image Processing
from PIL import Image
import skimage.io as skio

# Deep Learning
from sklearn.metrics import (confusion_matrix, ConfusionMatrixDisplay,
                             accuracy_score, precision_score,
                             recall_score, f1_score)
import torch
from torch.utils.data import DataLoader
from torch import nn, optim
import torchvision
from torchvision import datasets, transforms
from torchsummary import summary
from mtcnn import MTCNN
from facenet_pytorch import InceptionResnetV1

Let us also check if there is an available GPU for Deep Learning.

In [6]:
device = (torch.device('cuda') if torch.cuda.is_available()
          else torch.device('cpu'))
print(f"Your Device is {device}.")

Your Device is cuda.


<h2 style="color:#FDB3A5">Data Source</h2>

***

All Twice images used in this project are downloaded from the official website.<sup>[4]</sup> These images are owned by JYP Entertainment Co., Ltd. The images used are curated from all available Twice images on the website. There are two data folders used in this notebook: `raw_dataset.zip` and `group_image.zip` both of which can be downloaded from this google drive: 
* `raw_dataset.zip:` https://drive.google.com/file/d/19XYwiqSXk8YPJCRmxjlIhx3i2MR14H58/view?usp=sharing
* `group_image.zip:` https://drive.google.com/file/d/108_LuD1eIunF-O5Y4hHsMId7UxOsqqaC/view?usp=sharing

To use this dataset, first create a folder named `data` at the same level as your Python notebook. Move the downloaded `.zip` files inside the `data` folder and use the extract code.

In [7]:
# # Uncomment this code to extract the solo image dataset
# with zipfile.ZipFile(f"./data/raw_dataset.zip", 'r') as zip_ref:
#     zip_ref.extractall("./data/raw_dataset")

In [8]:
# # Uncomment this code to extract the group image dataset
# with zipfile.ZipFile(f"./data/group_image.zip", 'r') as zip_ref:
#     zip_ref.extractall("./data")

Now that we have extracted our dataset and set up all the necessary libraries, let us create our own Twice face recognition.

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Face Detection</h1>

<h2 style="color:#FDB3A5">MTCNN</h2>

***

We initialize the pre-trained model and try to use it first on a sample image to see the output.

In [28]:
# Load the pre-trained model
detector = MTCNN()

In [41]:
# Load sample image
img = skio.imread(f'./data/raw_dataset/dahyun/dahyun (10).jpg')

# Surpress the loading printout by the mtcnn library
stdout = sys.stdout
sys.stdout = open(os.devnull, 'w')

# Run the Pre-trained Model on our image and save the output
faces = detector.detect_faces(img)

# Restore stdout
sys.stdout.close()
sys.stdout = stdout

# Plot the image
fig, ax = plt.subplots()
ax.imshow(img)

# Loop for all the faces detected by the model
for face in faces:
    # Add the bounding box
    x, y, w, h = face['box']
    rectangle = Rectangle((x, y), w, h, fill=False, color='red')
    ax.add_patch(rectangle)

    # Add left eye patch
    le_x, le_y = face['keypoints']['left_eye']
    le_circ = Circle((le_x, le_y), 5, fill=True, color='red')
    ax.add_patch(le_circ)

    # Add right eye patch
    re_x, re_y = face['keypoints']['right_eye']
    re_circ = Circle((re_x, re_y), 5, fill=True, color='red')
    ax.add_patch(re_circ)

    # Add nose patch
    n_x, n_y = face['keypoints']['nose']
    n_circ = Circle((n_x, n_y), 5, fill=True, color='red')
    ax.add_patch(n_circ)

    # Add left mouth patch
    lm_x, lm_y = face['keypoints']['mouth_left']
    lm_circ = Circle((lm_x, lm_y), 5, fill=True, color='red')
    ax.add_patch(lm_circ)

    # Add right mouth patch
    rm_x, rm_y = face['keypoints']['mouth_right']
    rm_circ = Circle((rm_x, rm_y), 5, fill=True, color='red')
    ax.add_patch(rm_circ)

# Adds context to the image
ax.set_title('Face Detection using MTCNN')
ax.set_axis_off()
# plt.show()

# Save the figure as a PNG file
plt.savefig("./images/Sample_FaceDetection.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Sample_FaceDetection.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:40%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

# Display the output of the model . It contains a list of dictionaries
# Where each dictionary is a face the keys are:
# The bounding box, confidence, and keypoints (landmark positions)
print('\nOutput of MTCNN:')
print(faces[0])


Output of MTCNN:
{'box': [288, 153, 245, 313], 'confidence': 0.9992341995239258, 'keypoints': {'left_eye': (347, 281), 'right_eye': (460, 265), 'nose': (403, 342), 'mouth_left': (376, 398), 'mouth_right': (463, 386)}}


`MTCNN` outputs a dictionary containing the coordinates of the bounding box, the confidence of the model on its prediction, and the coordinates of the keypoints of the face (left_eye, right_eye, nose, mouth_left, mouth_right).

For our use case, we must create a cropped image based on the face's bounding box. This cropping will create an entire face image (no body or background). These cropped images are preferred for the *Face Classification* model later to have more accurate results.

<h2 style="color:#FDB3A5">Face Extraction</h2>

***

We will do the image cropping for all the images and then save them in a directory (`./data/cropped_images/`). The images will be automatically sorted to the correct folder indicating its name since we are using an already sorted dataset.

In [11]:
# Surpress the loading printout by the mtcnn library
stdout = sys.stdout
sys.stdout = open(os.devnull, 'w')

# Create directory for the cropped images
crop_dir = './data/cropped_images'
if os.path.exists(crop_dir):
    shutil.rmtree(crop_dir)
os.makedirs(crop_dir)

# Loop for all members of twice
names = ['chaeyoung', 'dahyun', 'jeongyeon', 'jihyo',
         'mina', 'momo', 'nayeon', 'sana', 'tzuyu']
for name in names:
    # Create directory for individual members
    twice_dir = crop_dir+f'/{name}'
    if os.path.exists(twice_dir):
        shutil.rmtree(twice_dir)
    os.makedirs(twice_dir)

    # Loops thorugh all 20 raw images for each member
    for i in range(1, 21):
        # Load current image
        img = skio.imread(f'./data/raw_dataset/{name}/{name} ({i}).jpg')

        # Run the Pre-trained Model on our image and save the output
        faces = detector.detect_faces(img)

        # Loop for all the faces detected by the model
        for idx, face in enumerate(faces):
            # Save the cropped image as defined by the bounding box
            x, y, w, h = face['box']
            skio.imsave(f"{twice_dir}/{name}_{i}.jpg",
                        img[y:y+h, x:x+w])

# Restore stdout
sys.stdout.close()
sys.stdout = stdout

By the end of this *Face Extraction* you will have a new folder inside your `data` folder called `cropped_images`. The new folder contains nine folders corresponding to the Twice members' names. Each folder contains 20 cropped images of the face of that member.

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Face Classification</h1>

<h2 style="color:#FDB3A5">Data Preprocessing</h2>

***

The first step in preprocessing is to set up the folder such that it splits the dataset into training (60% / 12 images), validation (20% / 4 images), and testing (20% / 4 images). The split data is in 3 separate folders, each containing a folder for each class. 

Let us also show a sample cropped image for each of the members.

In [12]:
# Copy labeled images to train, validation, and test directories
def create_dataset(src, dst, range_, class_):
    """Copy images of class class_ within range_ from src to dst.

    Parameters
    ----------
    src : str
        source directory
    dst : str
        destination directory
    range_ : tuple
        tuple of min and max image index to copy
    class_ : str
        image class 
    """
    # If existing, delete dir to reset
    if os.path.exists(dst):
        shutil.rmtree(dst)

    # Create destination directory
    os.makedirs(dst)

    # Copies the image from source to destination directory
    fnames = [f'{class_}_{i}.jpg' for i in range(*range_)]
    for fname in fnames:
        src_file = os.path.join(src, fname)
        dst_file = os.path.join(dst, fname)
        shutil.copyfile(src_file, dst_file)

In [13]:
# looping through create_dataset for each class
classes = ['chaeyoung', 'dahyun', 'jeongyeon', 'jihyo',
           'mina', 'momo', 'nayeon', 'sana', 'tzuyu']
for class_ in classes:
    # Source directory
    src = 'data/cropped_images'

    # Train directory
    dst = f'data/train_val_test/train/{class_}'
    create_dataset(src+'/'+class_, dst, range_=(9, 21), class_=class_)

    # Validation directory
    dst = f'data/train_val_test/validation/{class_}'
    create_dataset(src+'/'+class_, dst, range_=(1, 5), class_=class_)

    # Test directory
    dst = f'data/train_val_test/test/{class_}'
    create_dataset(src+'/'+class_, dst, range_=(5, 9), class_=class_)

In [42]:
# Load sample images for each class in the train set
chaeyoung = Image.open('data/train_val_test/train/chaeyoung/chaeyoung_10.jpg')
dahyun = Image.open('data/train_val_test/train/dahyun/dahyun_10.jpg')
jeongyeon = Image.open('data/train_val_test/train/jeongyeon/jeongyeon_10.jpg')
jihyo = Image.open('data/train_val_test/train/jihyo/jihyo_10.jpg')
mina = Image.open('data/train_val_test/train/mina/mina_10.jpg')
momo = Image.open('data/train_val_test/train/momo/momo_10.jpg')
nayeon = Image.open('data/train_val_test/train/nayeon/nayeon_10.jpg')
sana = Image.open('data/train_val_test/train/sana/sana_10.jpg')
tzuyu = Image.open('data/train_val_test/train/tzuyu/tzuyu_10.jpg')

# Plot the images
fig, ax = plt.subplots(1, 9, figsize=(30, 10))
ax[0].imshow(chaeyoung)
ax[0].set_title('Chaeyoung', fontsize=20)
ax[0].set_axis_off()
ax[1].imshow(dahyun)
ax[1].set_title('Dahyun', fontsize=20)
ax[1].set_axis_off()
ax[2].imshow(jeongyeon)
ax[2].set_title('Jeongyeon', fontsize=20)
ax[2].set_axis_off()
ax[3].imshow(jihyo)
ax[3].set_title('Jihyo', fontsize=20)
ax[3].set_axis_off()
ax[4].imshow(mina)
ax[4].set_title('Mina', fontsize=20)
ax[4].set_axis_off()
ax[5].imshow(momo)
ax[5].set_title('Momo', fontsize=20)
ax[5].set_axis_off()
ax[6].imshow(nayeon)
ax[6].set_title('Nayeon', fontsize=20)
ax[6].set_axis_off()
ax[7].imshow(sana)
ax[7].set_title('Sana', fontsize=20)
ax[7].set_axis_off()
ax[8].imshow(tzuyu)
ax[8].set_title('Tzuyu', fontsize=20)
ax[8].set_axis_off()
# plt.show()

# Save the figure as a PNG file
plt.savefig("./images/Sample_Faces.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Sample_Faces.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:100%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

We will also need to create dataloaders for our model. This will load the images and feed them to our model later. We will also introduce data augmentation (i.e., Random Horizontal Flip, Random Perspective, and Color Jitter) for a more robust model. The cropped images are also resized into 160 by 160 pixels since the pre-trained model (PTM) we will use is trained on such image size.

In [15]:
# Training set directory
train_dir = "data/train_val_test/train"

# Create data transformation to be use in calcultion of mean & std
data_transforms = transforms.Compose([
    transforms.Resize(size=(160, 160)),  # PTM are trained on 160 by 160 px
    transforms.ToTensor()
])
train_data = datasets.ImageFolder(root=train_dir, transform=data_transforms)

# Compute for the means and stds (for normalization)
imgs = torch.stack([img_t for img_t, _ in train_data], dim=3)
means = imgs.view(3, -1).mean(dim=1).numpy()
stds = imgs.view(3, -1).std(dim=1).numpy()

In [16]:
# Create data transformation for training, validation, and test sets
data_transforms = {
    'train': transforms.Compose([
        transforms.Resize(size=(160, 160)),  # PTM are trained on 160 by 160
        transforms.RandomHorizontalFlip(p=0.3),  # Data augmentation
        transforms.RandomPerspective(p=0.3),  # Data augmentation
        transforms.ColorJitter(brightness=0.3),  # Data augmentation
        transforms.ToTensor(),
        transforms.Normalize(means, stds)  # Normalize the image
    ]),
    'validation': transforms.Compose([
        transforms.Resize(size=(160, 160)),
        transforms.ToTensor(),
        transforms.Normalize(means, stds)
    ]),
    'test': transforms.Compose([
        transforms.Resize(size=(160, 160)),
        transforms.ToTensor(),
        transforms.Normalize(means, stds)
    ])
}

# Dataset directory
data_dir = 'data/train_val_test'

# loading image data using ImageFolder
image_datasets = {x: datasets.ImageFolder(os.path.join(data_dir, x),
                                          data_transforms[x])
                  for x in ['train', 'validation', 'test']}

# Dataloaders
dataloaders = {x: DataLoader(image_datasets[x], batch_size=36, shuffle=True)
               for x in ['train', 'validation', 'test']}

# Size of datasets
dataset_sizes = {x: len(image_datasets[x])
                 for x in ['train', 'validation', 'test']}

# Class names
class_names = image_datasets['train'].classes

<h2 style="color:#FDB3A5">Transfer Learning</h2>

***

Since our dataset is similar to the dataset used by the pre-trained model (PTM) and is small, we will use the PTM as a feature extractor. We will only replace the last dense layer with a more appropriate layer that outputs 9 values corresponding to our 9 classes.

In [17]:
# Load the PTM with its pre-trained weights base on facial recognition dataset
model = InceptionResnetV1(pretrained='vggface2')

# Must be set to True to be use in classification
model.classify = True

# Freeze the layers in our PTM
for param in model.parameters():
    param.requires_grad = False

The last layer we created flattens the output of the last convolutional layer then it uses a RelU activation function. A dropout was also added to prevent co-adaptation, and the final layer reduces the dimension to 9, corresponding to our 9 classes.

In [18]:
# Change the last layer named logits to the ff layer with 4 output channels
model.logits = nn.Sequential(nn.Linear(512*1*1, 512),
                             nn.ReLU(),
                             nn.Dropout(0.4),
                             nn.Linear(512, 9),
                             )

# Transfer the model to device (GPU) for faster training
model = model.to(device)

# set the loss function
criterion = nn.CrossEntropyLoss()

# only the parameters of the classifier are being optimized
optimizer = optim.Adam(model.logits.parameters(), lr=0.001)

<h2 style="color:#FDB3A5">Model Training</h2>

***

Now that we have modified our PTM, we can now proceed with the training. We will use 30 epochs in our training and get the model with the highest validation accuracy as our best model. We will also print the loss and accuracy through the training to track the progress.

In [19]:
def train_model(model, criterion, optimizer, num_epochs=25):
    """Function that trains the model using the given loss function,
    optimizer and number of epochs."""
    # Tracks training time
    since = time.time()

    # Saves the best model here later
    best_model_wts = copy.deepcopy(model.state_dict())
    best_acc = 0.0

    # Loop for all epoch
    for epoch in range(num_epochs):
        # Prints only every 5 epoch
        if epoch == 0 or epoch % 5 == 4:
            print(f'\nEpoch {epoch + 1}/{num_epochs}')
            print('-' * 10)

        # Each epoch has a training and validation phase
        for phase in ['train', 'validation']:
            if phase == 'train':
                model.train()  # Set model to training mode
            else:
                model.eval()   # Set model to evaluate mode

            # Tracker for loss
            running_loss = 0.0
            running_corrects = 0

            # Iterate over data.
            for inputs, labels in dataloaders[phase]:
                inputs = inputs.to(device)
                labels = labels.to(device)

                # Zero the parameter gradients
                optimizer.zero_grad()

                # Forward
                # Track history if only in train
                with torch.set_grad_enabled(phase == 'train'):
                    outputs = model(inputs)
                    _, preds = torch.max(outputs, 1)
                    loss = criterion(outputs, labels)

                    # Backward + optimize only if in training phase
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()

                # Statistics
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)

            # Calculate loss and accuracy of current epoch
            epoch_loss = running_loss / dataset_sizes[phase]
            epoch_acc = running_corrects.double() / dataset_sizes[phase]

            # Prints only every 4 epoch
            if epoch == 0 or epoch % 5 == 4:
                print(f'{phase} Loss: {epoch_loss:.4f} Acc: {epoch_acc:.4f}')

            # Deep copy the model if it beats the current best model
            if phase == 'validation' and epoch_acc > best_acc:
                best_acc = epoch_acc
                best_model_wts = copy.deepcopy(model.state_dict())

    # Print out model summary
    time_elapsed = time.time() - since
    print(f'\nTraining complete in '
          f'{time_elapsed // 60:.0f}m {time_elapsed % 60:.0f}s')
    print(f'Best val Acc: {best_acc:4f}')

    # load best model weights
    model.load_state_dict(best_model_wts)

    return model

In [20]:
# Train the model by calling the train_model function
model = train_model(model.cuda(),
                    criterion,
                    optimizer,
                    num_epochs=30)


Epoch 1/30
----------
train Loss: 2.0667 Acc: 0.2778
validation Loss: 1.6989 Acc: 0.6111

Epoch 5/30
----------
train Loss: 0.3555 Acc: 0.9630
validation Loss: 0.5202 Acc: 0.8889

Epoch 10/30
----------
train Loss: 0.0943 Acc: 0.9815
validation Loss: 0.3433 Acc: 0.8889

Epoch 15/30
----------
train Loss: 0.1046 Acc: 0.9722
validation Loss: 0.3625 Acc: 0.8889

Epoch 20/30
----------
train Loss: 0.1416 Acc: 0.9630
validation Loss: 0.3415 Acc: 0.9167

Epoch 25/30
----------
train Loss: 0.1432 Acc: 0.9537
validation Loss: 0.4700 Acc: 0.8889

Epoch 30/30
----------
train Loss: 0.0683 Acc: 0.9815
validation Loss: 0.4113 Acc: 0.8889

Training complete in 5m 44s
Best val Acc: 0.916667


<h2 style="color:#FDB3A5">Model Evaluation</h2>

***

For model evaluation, let us first calculate our dataset's Proportion Chance Criterion (PCC) to have a baseline comparison. Our dataset is balanced and consists of 180 faces, corresponding to the 20 faces of each 9 classes (`chaeyoung`, `dahyun`, `jeongyeon`, `jihyo`, `mina`, `momo`, `nayeon`, `sana`, `tzuyu`).

In [21]:
# Calculate the image count per class and the total image count
chaeyoung_cnt = len(os.listdir('./data/cropped_images/chaeyoung'))
dahyun_cnt = len(os.listdir('./data/cropped_images/dahyun'))
jeongyeon_cnt = len(os.listdir('./data/cropped_images/jeongyeon'))
jihyo_cnt = len(os.listdir('./data/cropped_images/jihyo'))
mina_cnt = len(os.listdir('./data/cropped_images/mina'))
momo_cnt = len(os.listdir('./data/cropped_images/momo'))
nayeon_cnt = len(os.listdir('./data/cropped_images/nayeon'))
sana_cnt = len(os.listdir('./data/cropped_images/sana'))
tzuyu_cnt = len(os.listdir('./data/cropped_images/tzuyu'))
total_cnt = (chaeyoung_cnt + dahyun_cnt + jeongyeon_cnt +
             jihyo_cnt + mina_cnt + momo_cnt +
             nayeon_cnt + sana_cnt + tzuyu_cnt)

# Calculate the PCC
pcc = ((chaeyoung_cnt/total_cnt)**2 +
       (dahyun_cnt/total_cnt)**2 +
       (jeongyeon_cnt/total_cnt)**2 +
       (jihyo_cnt/total_cnt)**2 +
       (mina_cnt/total_cnt)**2 +
       (momo_cnt/total_cnt)**2 +
       (nayeon_cnt/total_cnt)**2 +
       (sana_cnt/total_cnt)**2 +
       (tzuyu_cnt/total_cnt)**2
       )
print(f"     PCC = {pcc*100:.2f}%")
print(f"1.25*PCC = {1.25*pcc*100:.2f}%")

     PCC = 11.11%
1.25*PCC = 13.89%


The model has a PCC of `11.11%`. This means that 11.11% of the data can be classified correctly due to *chance* alone. With that as our baseline, let us evaluate our model by getting the predictions for our test set. After which, we will calculate the `accuracy`, `precision`, `recall`, and `f1-score`. The confusion matrix was also shown to determine which Twice member our model finds challenging to classify.

In [22]:
def evaluate(model, test_loader):
    """Function that test the model using the prepared test loader."""
    # Initialize counting of correct and total images
    correct = 0
    total = 0

    # Set no grad since this is evaluation mode
    with torch.no_grad():
        # Load the image and use the model to predict its label
        for imgs, labels in test_loader:
            imgs = imgs.to(device)
            labels = labels.to(device)
            outputs = model(imgs)
            _, predicted = torch.max(outputs, dim=1)
            total += labels.shape[0]
            correct += int((predicted == labels).sum())

    return predicted, labels

In [45]:
# Evaluate the model by calling the evaluate function
y_pred_tensor, y_true_tensor = evaluate(model, dataloaders['test'])
y_pred = y_pred_tensor.cpu().numpy()
y_true = y_true_tensor.cpu().numpy()

# Display Confusion Matrix
fig, ax = plt.subplots(1, 1, figsize=(5, 5))
cm = confusion_matrix(y_true, y_pred)
disp = ConfusionMatrixDisplay(cm)
disp.plot(ax=ax, cmap='RdPu')
disp.ax_.set_xticklabels(classes, fontsize=8)
disp.ax_.set_yticklabels(classes, fontsize=8)
disp.im_.colorbar.remove()
ax.tick_params(axis='x', rotation=45)
ax.set_title('Model Confusion Matrix')
# plt.show()

# Save the figure as a PNG file
plt.tight_layout()
plt.savefig("./images/Confusion_Matrix.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Confusion_Matrix.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:30%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

# Print Evaluation scores
print('Model Evaluation:')
print(f'Accuracy Score: {accuracy_score(y_true, y_pred)*100:.2f}')
print(f'Precision Score: '
      f'{precision_score(y_true, y_pred, average="macro")*100:.2f}')
print(f'Recall Score: '
      f'{recall_score(y_true, y_pred, average="macro")*100:.2f}')
print(f'F1 Score: {f1_score(y_true, y_pred,  average="macro")*100:.2f}')

Model Evaluation:
Accuracy Score: 91.67
Precision Score: 92.22
Recall Score: 91.67
F1 Score: 91.62


The model has `91.67%` accuracy, much higher than the calculated PCC. This means that our model is a successful classification since we were able to beat the baseline. We also got a high precision, recall, and F1 score. 

We could identify which member finds it difficult to classify in the confusion matrix. Here `Tzuyu` is misclassified as `Jeongyeon`, `Mina` is misclassified as `Momo`, and `Momo` is misclassified as `Tzuyu`. These are the members others might find difficult to remember their names and faces.

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Twice Face Recognition</h1>

Now that we have a trained model for identifying the members of Twice let us put it to a test. A group image of Twice will be tested, and let us see if the model can correctly associate the faces with the correct names.

In [46]:
# Load the pre-trained model
detector = MTCNN()

# Load sample image
img = skio.imread(f"./data/group_image/twice_0.jpg")

# Surpress the loading printout by the mtcnn library
stdout = sys.stdout
sys.stdout = open(os.devnull, 'w')

# Saves the output of the model (including bounding box)
faces = detector.detect_faces(img)

# Restore stdout
sys.stdout.close()
sys.stdout = stdout

# If existing, delete dir to reset
face_recog_dir = 'data/face_recog'
if os.path.exists(face_recog_dir):
    shutil.rmtree(face_recog_dir)

# Create destination directory
os.makedirs(f"{face_recog_dir}/test/dahyun")

# Loop for all the faces detected by the model
for idx, face in enumerate(faces):
    # Save the cropped image as defined by the bounding box
    x, y, w, h = face['box']
    skio.imsave(f"./{face_recog_dir}/test/dahyun/sample_{idx}.jpg",
                img[y:y+h, x:x+w])

# loading cropped image data using ImageFolder
face_datasets = datasets.ImageFolder(os.path.join(face_recog_dir, 'test'),
                                     data_transforms['test'])

# Cropped Image Dataloaders
face_dataloaders = DataLoader(face_datasets, batch_size=9, shuffle=False)

# Get batch of images from the test DataLoader
images, _ = next(iter(face_dataloaders))
images = images.to(device)

# Predicts the images using the trained model
outputs = model(images)
_, predicted = torch.max(outputs, 1)

# Gets the class labels from the predictions
names = [classes[i] for i in predicted]

# Plot the image
fig, ax = plt.subplots(figsize=(10, 10))
ax.imshow(img)

# Loop for all the faces detected by the model
for idx, face in enumerate(faces):
    # Add the bounding box
    x, y, w, h = face['box']
    rectangle = Rectangle((x, y), w, h, fill=False, color='#FE7BA3')
    ax.add_patch(rectangle)

    # Add the face labels
    bbox_props = dict(boxstyle="round,pad=0.2", fc="#FE7BA3", ec="#FE7BA3")
    ax.text(x+3, y-5, names[idx], fontsize=7, color='white', bbox=bbox_props)

# Adds context to the image
ax.set_title(f'\nFace Recognition of Twice')
ax.set_axis_off()
# plt.show()

# Save the figure as a PNG file
plt.savefig("./images/Twice_Recognition.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Twice_Recognition.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:70%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

Even at `91.67%` accuracy, we could correctly predict the sample group image. To further boost our model's performance, adding more images would be beneficial. But at its current state, our model is enough to help the new Once in recognizing each Twice member.

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">Key Takeaways</h1>

Deep Learning methods can be applied even in the use-case of being a K-pop fan. We demonstrated in this project that:
* There are available pre-trained model to do *Face Detection* and *Face Classification*
* Even with a small dataset (12 training images and 4 validation images), we can tune (by Feature Extraction) the model toward our specific purpose
* Analyzing the confusion matrix, we could determine which member Twice, people might find it difficult to associate the name with the face

Now you know how to do Face Recognition using the Deep Learning method. You can simply swap out the dataset to another K-pop group you like and make a Face Recognition model yourself. 

A great question to ask now is, Are you already a Once? And also, did you catch who is my bias in Twice?

<h1 style="background-color:#FE7BA3; color:#F5F5F1; padding: 10px;">References</h1>

[1] K. Zhang, Z. Zhang, Z. Li and Y. Qiao. (2016). *Joint Face Detection and Alignment Using Multitask Cascaded Convolutional Networks*, IEEE Signal Processing Letters.

[2] Szegedy, C., Ioffe, S., Vanhoucke, V., & Alemi, A. (2016). *Inception-v4, Inception-ResNet and the Impact of Residual Connections on Learning*. In Proceedings of the Computer Vision and Pattern Recognition (CVPR) (pp. 4278-4284).

[3] Q. Cao, L. Shen, W. Xie, O. M. Parkhi, A. Zisserman (2018). *VGGFace2: A dataset for recognising faces across pose and age*. International Conference on Automatic Face and Gesture Recognition.

[4] JYP Entertainment Co., Ltd (n.d.) *Twice: Gallery*. Retrieved June 11, 2023 from https://twice.jype.com/Default/Gallery

[5] OpenAI. (2023). *ChatGPT* (Mar 14 version) [Large language model]